soccer_ipm/soccer_ipm/msgs/markings.py

```diff
@@ -0,0 +1,243 @@
+        logger: RcutilsLogger) -> List[sv3dm.MarkingSegment]:
+    """
+    Map a given list of 2D field marking segments onto the field plane.
+
+    :param header: Message header of the original detection
+    :param intersections_2d: The 2D message that should be mapped
+    :param ipm: An instance of the IPM mapping utility
+    :param output_frame: The tf frame of the field
+    :param logger: A ros logger to display warnings etc.
+    :returns: The segments as 3D cartesian detections in the output_frame
+    """
+    field = create_field_plane(header.stamp, output_frame)
+
+    # Convert segment start and end points to np array
+    segment_points_np = np.array([(
+        (segment.start.x, segment.start.y),
+        (segment.end.x, segment.end.y)) for segment in marking_segments_2d])
+
+    # Map all points at once from image onto field plane
+    segments_on_plane = ipm.map_points(
+        field,
+        segment_points_np.reshape(-1, 2),
+        header,
+        output_frame=output_frame).reshape(-1, 2, 3)
+
+    # Convert the numpy array back to the soccer vision messages datatype
+    marking_segments_3d = []
+    for i, segment in enumerate(segments_on_plane):
+        # Check if any of the points failed to map
+        if np.any(np.isnan(segment)):
+            logger.warn(
+                'Got a segment I could not transform.',
+                throttle_duration_sec=5)
+            continue
+        start, end = segment
+        segment_msg = sv3dm.MarkingSegment()
+        # Get confidence value from original list
+        segment_msg.confidence.confidence = marking_segments_2d[i].confidence.confidence
+        # Convert np array -> Point
+        segment_msg.start.x = start[0]
+        segment_msg.start.y = start[1]
+        segment_msg.start.z = start[2]
+        segment_msg.end.x = end[0]
+        segment_msg.end.y = end[1]
+        segment_msg.end.z = end[2]
+        marking_segments_3d.append(segment_msg)
+    return marking_segments_3d
+
+
+def map_marking_ellipses(
+        header: Header,
+        ellipses_2d: List[sv2dm.MarkingEllipse],
+        ipm: IPM,
+        output_frame: str,
+        logger: RcutilsLogger) -> List[sv3dm.MarkingEllipse]:
+    """
+    Map a given list of 2D field marking ellipses onto the field plane.
+
+    :param header: Message header of the original detection
+    :param intersections_2d: The 2D message that should be mapped
+    :param ipm: An instance of the IPM mapping utility
+    :param output_frame: The tf frame of the field
+    :param logger: A ros logger to display warnings etc.
+    :returns: The ellipses as 3D cartesian detections in the output_frame
+    """
+    field = create_field_plane(header.stamp, output_frame)
+
+    ellipses_3d = []
+    ellipse: sv2dm.MarkingEllipse
+    for ellipse in ellipses_2d:
+        # Create center point
+        center = Point2DStamped(
+            header=header,
+            point=ellipse.center
+        )
+
+        # TODO check math
```

Comment:
Have you checked? :stuck_out_tongue_closed_eyes:

Reply:
I did ^^ At least I think so :D